x-pack/plugins/apm/server/lib/services/get_service_dependencies/index.ts

```diff
@@ -5,226 +5,315 @@
+              : null,
+        })),
+      },
+      errorRate: {
+        value:
+          mergedMetrics.value.count > 0
+            ? (mergedMetrics.value.error_count ?? 0) / mergedMetrics.value.count
+            : null,
+        timeseries: mergedMetrics.timeseries.map((point) => ({
+          x: point.x,
+          y: point.count > 0 ? (point.error_count ?? 0) / point.count : null,
+        })),
+      },
+    };
+
+    if (item.service) {
+      return {
+        name: item.service.name,
+        type: 'service' as const,
+        serviceName: item.service.name,
+        environment: item.service.environment,
+        // agent.name should always be there, type returned from joinByKey is too pessimistic
+        agentName: item.agent!.name,
+        ...destMetrics,
+      };
+    }
+
+    return {
+      name: item.span.destination.service.resource,
+      type: 'external' as const,
+      spanType: item.span.type,
+      spanSubtype: item.span.subtype,
+      ...destMetrics,
+    };
+  });
+
+  const latencySums = metricsByResolvedAddress
+    .map(
+      (metric) => (metric.latency.value ?? 0) * (metric.throughput.value ?? 0)
+    )
+    .filter(isFiniteNumber);
+
+  const minLatencySum = Math.min(...latencySums);
+  const maxLatencySum = Math.max(...latencySums);
+
+  return metricsByResolvedAddress.map((metric) => {
+    const impact =
+      isFiniteNumber(metric.latency.value) &&
+      isFiniteNumber(metric.throughput.value)
+        ? ((metric.latency.value * metric.throughput.value - minLatencySum) /
+            (maxLatencySum - minLatencySum)) *
+          100
+        : 0;
+
+    return {
+      ...metric,
+      impact,
+    };
+  });
+}
+
+async function getServiceMetricsAndConnections({
+  setup,
+  serviceName,
+  environment,
+  numBuckets,
+  start,
+  end,
+}: {
+  serviceName: string;
+  setup: Setup;
+  environment?: string;
+  numBuckets: number;
+  start: number;
+  end: number;
+}) {
+  return await Promise.all([
+    getMetrics({
+      setup,
+      serviceName,
+      environment,
+      numBuckets,
+      start,
+      end,
+    }),
+    getConnections({
+      setup,
+      serviceName,
+      environment,
+      start,
+      end,
+    }),
+  ]);
+}
 
-export function getServiceDependencies({
+export async function getServiceDependenciesPerPeriod({
   setup,
   serviceName,
   environment,
   numBuckets,
+  comparisonStart,
+  comparisonEnd,
 }: {
   serviceName: string;
   setup: Setup & SetupTimeRange;
   environment?: string;
   numBuckets: number;
-}): Promise<ServiceDependencyItem[]> {
+  comparisonStart?: number;
+  comparisonEnd?: number;
+}) {
   return withApmSpan('get_service_dependencies', async () => {
     const { start, end } = setup;
-    const [allMetrics, destinationMap] = await Promise.all([
-      getMetrics({
+
+    const [
+      [currentPeriodMetrics, currentPeriodConnections],
+      [previousPeriodMetrics, previousPeriodConnections] = [],
```

Comment:
Any reason why this is an array and not a key-value map?

Reply:
not really, only because `getServiceMetricsAndConnections` returns a `Promise.all` which returns an array. I can change it to return a key-value map if you think it's important though.

Reply:
Yeah let's do that - I don't think double array destructuring is very legible (to me it's the same category as >3 function arguments)

Reply:
np 👍🏻